### PR TITLE
PRP-13 Add Database

### PR DESCRIPTION
**Pull request by aplybeah**
## Ticket

https://wicmtdp.atlassian.net/browse/PRP-13

## Changes
* Add Aurora Serverless DB
* add staging module
TODO:
* randomly generate master pw and add to parameter store
* figure out adding perms via [tagging](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/security_iam_id-based-policy-examples.html#security_iam_id-based-policy-examples-create-db-instance-specified-groups) for the database
* checkov backup compliance

## Testing
1. Log into AWS
2. Navigate to `RDS` and make sure the availability zone is `us-west (N. California)`
3. Check that a DB instance exists
4. Check that connected resources (e.g. security groups) exist

### CODE REVIEW

File: infra/modules/service/main.tf [hunk 1]
```diff
@@ -319,3 +319,28 @@ resource "aws_security_group" "app" {
...
+###########################
+## Database Configuration ##
+###########################
+resource "aws_rds_cluster" "postgresql" {
+  cluster_identifier = "${var.service_name}"
+  engine = "aurora-postgresql"
+  engine_mode = "provisioned"
+  database_name = replace("${var.service_name}", "-", "_")
+  master_username = "app_usr"
+  master_password = "test12345"
```
Comment (microwavenby) [inline infra/modules/service/main.tf]: Even for a test, we probably shouldn't use a password hardcoded into the terraform

I have this example from my own terraform that uses secrets manager, and i believe similar is possible with parameter store (secure strings)
```terraform
  password                = data.aws_secretsmanager_secret_version.rds_admin_password.secret_string
```


```terraform
resource "random_password" "rds_admin_password_value"{
  length           = 24
  special          = true
  override_special = "_!%^"
}

resource "aws_secretsmanager_secret" "rds_admin_password" {
  name = "metadata/rds/admin-password"
}

resource "aws_secretsmanager_secret_version" "rds_admin_password_version" {
  secret_id = aws_secretsmanager_secret.rds_admin_password.id
  secret_string = random_password.rds_admin_password_value.result
}

```

Comment (aplybeah) [reply infra/modules/service/main.tf]: That's fairly similar to what we did on the demo.

File: infra/modules/service/main.tf [hunk 2]
```diff
@@ -343,4 +343,20 @@ resource "aws_rds_cluster_instance" "postgresql-cluster" {
...
+  override_special = "!#$%&*()-_=+[]{}<>:?"
+}
+
+resource "aws_secretsmanager_secret" "random_db_password" {
+  name  = "metadata/db/admin-password"
+}
+
+resource "aws_secretsmanager_secret_version" "random_db_password" {
+  secret_id = aws_secretsmanager_secret.random_db_password.id
+  secret_string = random_password.random_db_password.result
```
Comment (microwavenby) [inline infra/modules/service/main.tf]:
```suggestion


resource "aws_ssm_parameter" "admin_db_password" {
  name = "/metadata/db/admin-password"
  description = "Admin database password"
  type = "SecureString"
  value = random_password.random_db_password.result

  tags = {
    environment = "staging"
  }


```

File: infra/modules/service/main.tf [hunk 3]
```diff
@@ -329,7 +329,7 @@ resource "aws_rds_cluster" "postgresql" {
   engine_mode = "provisioned"
   database_name = replace("${var.service_name}", "-", "_")
   master_username = "app_usr"
-  master_password = "test12345"
+  master_password = "${aws_secretsmanager_secret_version.random_db_password.secret_string}"
```
Comment (microwavenby) [inline infra/modules/service/main.tf]:
```suggestion
  master_password = "${aws_ssm_parameter.admin_db_password.secret_string}"
```

**Review by microwavenby — COMMENTED** [review]
Not to ask you to change it twice, but I only alluded to not wanting to use secretsmanager again..

Secretsmanager has some protections against accidental deletion that make it fairly cumbersome to use, especially for a short term demonstration.

I think it'd be thoroughly suitable to use SSM's secure string for this.

**Review by microwavenby — APPROVED** [review]

File: infra/modules/service/main.tf [hunk 4]
```diff
@@ -319,3 +319,111 @@ resource "aws_security_group" "app" {
...
+resource "aws_backup_plan" "postgresql" {
+  name = "${var.service_name}_backup_plan"
+
+  rule {
+    rule_name         = "${var.service_name}_backup_rule"
+    target_vault_name = "${var.service_name}-vault"
+    schedule          = "cron(0 12 ? * SUN *)"
+  }
+}
+# checkov:skip=CKV2_AWS_8:TODO add backup selection plan using tags
```
Comment (rocketnova) [inline infra/modules/service/main.tf]: Looks like this checkov is still failing. Including a backup selection as part of this PR should probably fix that

File: infra/modules/service/main.tf [hunk 5]
```diff
@@ -319,3 +319,111 @@ resource "aws_security_group" "app" {
...
+      type        = "Service"
+      identifiers = ["monitoring.rds.amazonaws.com"]
+    }
+  }
+}
+
+################################################################################
+# Parameters for Query Logging
+################################################################################
+# checkov:skip=CKV2_AWS_27:have concerns about sensitive data in logs; want better way to get this information
```
Comment (rocketnova) [inline infra/modules/service/main.tf]: @aplybeah I agree with this comment to skip query logging for now. Can you open a ticket so we can talk through db logging as a team? Also, I'm not sure this checkov skip is working?

**Review by rocketnova — COMMENTED** [review]
Hmm looks like checkov still isn't passing because we're missing backup selection

**Pull request merged**